presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java

```diff
@@ -313,22 +321,390 @@ public void testHardCpuLimit()
+        Stream.of(root, child).forEach(group -> assertExceedsCpuLimit(group, 4));
+
+        // q2 gets queued, because the cached usage is greater than the limit
+        MockManagedQueryExecution q2 = new MockManagedQueryExecutionBuilder().build();
+        child.run(q2);
+        assertEquals(q2.getState(), QUEUED);
+
+        // Generating CPU quota before the query finishes. This assertion is useful for testing CPU update during quota generation,
+        // without the invocation of updateGroupsAndProcessQueuedQueries.
+        root.generateCpuQuota(2);
+        Stream.of(root, child).forEach(group -> assertWithinCpuLimit(group, 2));
+
+        // An incoming query starts running right away.
+        MockManagedQueryExecution q3 = new MockManagedQueryExecutionBuilder().build();
+        child.run(q3);
+        assertEquals(q3.getState(), RUNNING);
+
+        // A queued query starts running only after invoking `updateGroupsAndProcessQueuedQueries`.
+        assertEquals(q2.getState(), QUEUED);
+        root.updateGroupsAndProcessQueuedQueries();
+        assertEquals(q2.getState(), RUNNING);
+    }
+
+    @Test(timeOut = 10_000)
+    public void testCpuUsageUpdateAtQueryCompletion()
+    {
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup child = root.getOrCreateSubGroup("child");
+
+        // Set the common configurations for the two groups
+        Stream.of(root, child).forEach(group -> {
+            group.setCpuQuotaGenerationMillisPerSecond((new Duration(1, MILLISECONDS)).toMillis());
+            group.setHardCpuLimit(new Duration(3, MILLISECONDS));
+            group.setSoftCpuLimit(new Duration(3, MILLISECONDS));
+            group.setHardConcurrencyLimit(100);
+            group.setMaxQueuedQueries(100);
+        });
+
+        MockManagedQueryExecution q1 = new MockManagedQueryExecutionBuilder().build();
+        child.run(q1);
+        assertEquals(q1.getState(), RUNNING);
+
+        q1.consumeCpuTime(new Duration(4, MILLISECONDS));
+        q1.complete();
+        // Query completion updates the cached usage to 2s. This is important because q1 will be removed
+        // from runningQueries at this point. Therefore, updateGroupsAndProcessQueuedQueries invocation will not be
+        // able to update CPU usage for it later.
+        Stream.of(root, child).forEach(group -> assertExceedsCpuLimit(group, 4));
+
+        // q2 gets queued since cached usage exceeds the limit.
+        MockManagedQueryExecution q2 = new MockManagedQueryExecutionBuilder().build();
+        child.run(q2);
+        assertEquals(q2.getState(), QUEUED);
+
+        root.generateCpuQuota(2);
+        Stream.of(root, child).forEach(group -> assertWithinCpuLimit(group, 2));
+        assertEquals(q2.getState(), QUEUED);
+
+        // q2 starts running after groups are updated. The assertion is to make sure that CPU usage is not double counted.
```

Comment:
Nit: These comments are helpful but could be less verbose. For example: `q2 should run after groups are updated. CPU usage should not be double counted`